# Homework 3

#Question 1
Python users, use Python's pandas library to read in the Stata-formatted dataset used in class called "SI Sales.dta".  R users, use R's foreign library to read in the Stata-formatted dataset called "SI Sales Old.dta".  Replicate all of the regression results for this dataset that I presented in class.

In [1]:
%pylab inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import Quandl
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.graphics.api import abline_plot
import patsy
import seaborn as sns
sns.set(context='notebook', style='whitegrid', palette='deep', font='sans-serif', font_scale=1, rc=None)
import urllib2 as url

Populating the interactive namespace from numpy and matplotlib


In [2]:
sales = pd.read_stata('SI Sales.dta')
print sales.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31680 entries, 0 to 31679
Data columns (total 6 columns):
price         31680 non-null float64
unit_size     31680 non-null float32
land_size     31680 non-null float32
age           31680 non-null float32
todt          31680 non-null float32
sales_year    31680 non-null float32
dtypes: float32(5), float64(1)
memory usage: 1.1 MB
None


In [3]:
print sales.describe()

                price     unit_size     land_size           age          todt  \
count    31680.000000  31680.000000  31680.000000  31680.000000  31680.000000   
mean    402106.300221    161.038956    326.063385     35.846844      0.007102   
std     239551.524308     79.404528    289.825500     31.082471      0.083968   
min      50500.000000     20.067047     24.340586      0.000000      0.000000   
25%     291975.000000    114.456497    185.806000      9.000000      0.000000   
50%     371000.000000    145.671906    252.696167     27.000000      0.000000   
75%     455000.000000    184.969879    380.902313     56.000000      0.000000   
max    7700000.000000   1431.077759  10788.825195    119.000000      1.000000   

         sales_year  
count  31680.000000  
mean    2006.567139  
std        2.766369  
min     2003.000000  
25%     2004.000000  
50%     2006.000000  
75%     2009.000000  
max     2013.000000  


In [4]:
print sales.corr()

               price  unit_size  land_size       age      todt  sales_year
price       1.000000   0.549706   0.505356 -0.064088  0.352151    0.096092
unit_size   0.549706   1.000000   0.446105 -0.201536  0.326307    0.023416
land_size   0.505356   0.446105   1.000000  0.274697  0.346801    0.077402
age        -0.064088  -0.201536   0.274697  1.000000  0.005157    0.125040
todt        0.352151   0.326307   0.346801  0.005157  1.000000    0.012260
sales_year  0.096092   0.023416   0.077402  0.125040  0.012260    1.000000


In [5]:
mod = smf.ols(formula='price ~ unit_size', data = sales).fit()
print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.302
Model:                            OLS   Adj. R-squared:                  0.302
Method:                 Least Squares   F-statistic:                 1.372e+04
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:31   Log-Likelihood:            -4.3166e+05
No. Observations:               31680   AIC:                         8.633e+05
Df Residuals:                   31678   BIC:                         8.633e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept    1.35e+05   2542.335     53.118      0.0

In [6]:
mod.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                    Results: Ordinary least squares
=======================================================================
Model:               OLS               Adj. R-squared:      0.302      
Dependent Variable:  price             AIC:                 863318.9733
Date:                2015-10-05 02:52  BIC:                 863335.7002
No. Observations:    31680             Log-Likelihood:      -4.3166e+05
Df Model:            1                 F-statistic:         1.372e+04  
Df Residuals:        31678             Prob (F-statistic):  0.00       
R-squared:           0.302             Scale:               4.0046e+10 
-----------------------------------------------------------------------
             Coef.     Std.Err.    t     P>|t|     [0.025      0.975]  
-----------------------------------------------------------------------
Intercept 135043.8679 2542.3346  53.1181 0.0000 130060.7932 140026.9425
unit_size   1658.3706   14.1594 117.1215 0.0000   1630.6176   1686.1236
-----------------------------------------------------------------------
Omnibus:            51914.230      Durbin-Watson:         1.973        
Prob(Omnibus):      0.000          Jarque-Bera (JB):      103733692.864
Skew:               10.534         Prob(JB):              0.000        
Kurtosis:           282.539        Condition No.:         406          
=======================================================================

"""

In [7]:
mod = smf.ols(formula='price ~ unit_size + land_size', data = sales).fit()
print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.387
Model:                            OLS   Adj. R-squared:                  0.387
Method:                 Least Squares   F-statistic:                     9985.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:            -4.2961e+05
No. Observations:               31680   AIC:                         8.592e+05
Df Residuals:                   31677   BIC:                         8.593e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept   1.179e+05   2397.598     49.177      0.0

In [8]:
mod = smf.ols(formula='price ~ unit_size + land_size + age', data = sales).fit()
print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.393
Model:                            OLS   Adj. R-squared:                  0.393
Method:                 Least Squares   F-statistic:                     6840.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:            -4.2945e+05
No. Observations:               31680   AIC:                         8.589e+05
Df Residuals:                   31676   BIC:                         8.589e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept   1.496e+05   2940.848     50.854      0.0

In [9]:
mod = smf.ols(formula='price ~ unit_size + land_size + age + todt', data = sales).fit()
print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.406
Model:                            OLS   Adj. R-squared:                  0.406
Method:                 Least Squares   F-statistic:                     5421.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:            -4.2910e+05
No. Observations:               31680   AIC:                         8.582e+05
Df Residuals:                   31675   BIC:                         8.582e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept   1.674e+05   2985.210     56.076      0.0

In [10]:
sales['priceper1000'] = sales['price']/1000
print sales.head()

    price   unit_size   land_size  age  todt  sales_year  priceper1000
0  327500  142.699005  232.257507   32     0        2011       327.500
1  346314  195.096298  239.318130   10     0        2006       346.314
2  349830  179.488602  201.413712    1     0        2006       349.830
3  325000  174.750549  227.612350  106     0        2005       325.000
4  285000   97.083633  271.741272  104     0        2003       285.000


In [11]:
mod = smf.ols(formula='priceper1000 ~ unit_size + land_size + age + todt', data = sales).fit()
print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:           priceper1000   R-squared:                       0.406
Model:                            OLS   Adj. R-squared:                  0.406
Method:                 Least Squares   F-statistic:                     5421.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:            -2.1026e+05
No. Observations:               31680   AIC:                         4.205e+05
Df Residuals:                   31675   BIC:                         4.206e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept    167.3977      2.985     56.076      0.0

In [12]:
sales['logprice'] = np.log(sales['price'])
sales['logunit'] = np.log(sales['unit_size'])
sales['logland'] = np.log(sales['land_size'])
sales['logage'] = np.log(sales['age']+1)
mod = smf.ols(formula='logprice ~ logunit + logland + logage + todt', data = sales).fit()
print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:               logprice   R-squared:                       0.364
Model:                            OLS   Adj. R-squared:                  0.364
Method:                 Least Squares   F-statistic:                     4540.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:                -12511.
No. Observations:               31680   AIC:                         2.503e+04
Df Residuals:                   31675   BIC:                         2.507e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      9.7769      0.029    332.252      0.0

# Question 2
There is an additional feature in this dataset called "sales_year", which captures the year the sale of a house in Staten Island occurred.  From this feature, generate a feature that is linear time trend.  (A linear time trend is a feature that takes on value "1" in the initial year and increments by "1" each subsequent year.  For example, if 2003 were "1", 2004 would be "2", 2005 would be "3", and so forth.)  Run a linear regression model that relates the sales price to unit size, land size, age, the Todt Hill indicator, and the linear time trend.  How would you interpret the estimated coefficient associated with the linear time trend?  What is the 95% confidence interval of your interpretation?  Based on your regression diagnostics, have you improved the fit of the house price sales data by including the linear time trend as an additional explanatory feature?  

In [13]:
sales['yearSale'] = sales['sales_year']-2002
mod = smf.ols(formula='price ~ unit_size + land_size + age + todt + yearSale', data = sales).fit()
print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.412
Model:                            OLS   Adj. R-squared:                  0.412
Method:                 Least Squares   F-statistic:                     4431.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:            -4.2896e+05
No. Observations:               31680   AIC:                         8.579e+05
Df Residuals:                   31674   BIC:                         8.580e+05
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept   1.435e+05   3295.435     43.547      0.0

## Comment
###An increase in linear time trend of one year increases the sales price by ~$6325. The 95% Confidence Interval of linear time trend is [5586.892, 7064.488], which is not overlapped with confidence intervals of other parameters. We can observe that the R-squared value increases from 0.364 to 0.412, indicating that we have improved the fit of house price by including the linear time trend as an additional explanatory feature.

#Question 3
As noted in class, the unit size and land size features are measured in squared meters. Suppose I ask you to re-express these features using the Imperial system of square feet rather than square meters, but I express a concern that the interpretation of the estimated coefficients, such as age, would be changed. Without acutally doing any statistical learning, what would you say to me about my concern? Rerun the linear regression in 2. using the dwelling size and land size measured in square feet (rather than square meters). What, if anything, has changed in your estimated coefficients?

In [14]:
sales['unitSize'] = sales['unit_size']*10.7639
sales['landSize'] = sales['land_size']*10.7639
mod = smf.ols(formula='price ~ unitSize + landSize + age + todt + yearSale', data = sales).fit()
print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.412
Model:                            OLS   Adj. R-squared:                  0.412
Method:                 Least Squares   F-statistic:                     4431.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:            -4.2896e+05
No. Observations:               31680   AIC:                         8.579e+05
Df Residuals:                   31674   BIC:                         8.580e+05
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept   1.435e+05   3295.435     43.547      0.0

## Comment
###By comparing the summary table above with the result of question #2, we can conclude that changing unit of certain parameters (unit size and land size) will not affect the estimates of other parameters (age). Also, the coefficient of unit size and land size was increased by a factor of 10.7639, which is the same as the conversion from square meter to square feet.

# Question 4
Assume the following data generating process (DGP) governs a random sample of size 10,000: yi=β0+β1x1i+β2x2i+ϵi for ϵi∼N(0,1). Further assume for this DGP that β0=β1=β2=1. (a) Suppose the following process governs your features: x1i∼N(0,1) and x2i∼N(0,1) are independent. Using R or Python, calculate the correlation between features x1i and x2i. Mistakenly, you decide to estimate a linear regression that includes only the feature x1i. Using R or Python, simulate this DGP and run the mistaken linear regression that includes only feature x1i. What value do you obtain for the coefficient associated with with feature x1i? (b) Suppose instead that the follow process governs your features: x1i=zi+ηi and x2i=−zi+ωi, where zi∼N(0,1), ηi∼N(0,1), and ωi∼N(0,1) are independent. Using R or Python, calculate the correlation between features x1i and x2i. Again, you mistakenly decide to estimate a linear regression that includes only the feature x1i. Using R or Python, simulate this DGP and run the mistaken linear regression that includes only feature x1i. What value do you obtain for the coefficient associated with with feature x1i? (c) Are there any conclusions you can draw from your results in (a) and (b)?

In [15]:
e = np.random.normal(0, 1, 10000)
x1 = np.random.normal(0, 1, 10000)
x2 = np.random.normal(0, 1, 10000)
y1 = 1 + x1 + x2 + e
x1f = pd.DataFrame(x1)
x2f = pd.DataFrame(x2)
y1f = pd.DataFrame(y1)
data1 = pd.merge(y1f, x1f, left_index=True, right_index=True)
mod1 = smf.ols(formula='y1 ~ x1', data=data1).fit()
print(mod1.summary())
data2 = pd.merge(x1f, x2f, left_index=True, right_index=True)
print data2.corr()

                            OLS Regression Results                            
Dep. Variable:                     y1   R-squared:                       0.334
Model:                            OLS   Adj. R-squared:                  0.334
Method:                 Least Squares   F-statistic:                     5023.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:                -17679.
No. Observations:               10000   AIC:                         3.536e+04
Df Residuals:                    9998   BIC:                         3.538e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      1.0201      0.014     71.954      0.0

## Comment
###We noticed that the correlation between x1 and x2 is about 0, indicating that x1 and x2 are independent. In the linear regression includes only feature x1, the coefficient associated with x1 is 1.0185, which is very close to the parameter in our model (beta1 = 1).

In [16]:
e1 = np.random.normal(0, 1, 10000)
z = np.random.normal(0, 1, 10000)
n = np.random.normal(0, 1, 10000)
w = np.random.normal(0, 1, 10000)
x12 = z + n
x22 = -z + w
y2 = 1 + x12 + x22 + e1
x12f = pd.DataFrame(x12)
x22f = pd.DataFrame(x22)
y2f = pd.DataFrame(y2)
data3 = pd.merge(y2f, x12f, left_index=True, right_index=True)
mod2 = smf.ols(formula='y2 ~ x12', data=data3).fit()
print(mod2.summary())
data4 = pd.merge(x12f, x22f, left_index=True, right_index=True)
print data4.corr()

                            OLS Regression Results                            
Dep. Variable:                     y2   R-squared:                       0.161
Model:                            OLS   Adj. R-squared:                  0.161
Method:                 Least Squares   F-statistic:                     1913.
Date:                Mon, 05 Oct 2015   Prob (F-statistic):               0.00
Time:                        02:52:32   Log-Likelihood:                -18904.
No. Observations:               10000   AIC:                         3.781e+04
Df Residuals:                    9998   BIC:                         3.783e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      1.0061      0.016     62.783      0.0

## Comment
###In case (b), the parameter z is a component of both x1 and x2, making x1 and x2 not indepent anymore. This will affect the ultimate fit of the original model. In the linear regression includes only feature x1, the coefficient associated with x1 is 0.5004, which is not close to the parameter in our model (beta1 = 1). Also, the value of R-squared is smaller than the previous model, indicating that the second fit is not as good as the first one.